File: CppMMO/GameServer/QuestComponent.cpp
#include "pch.h"
#include "QuestComponent.h"

void QuestComponent::Update()
{
	std::cout << "QuestComponent is updating!" << std::endl;
}

void QuestComponent::Notify(const EventData& message)
{
	NotifyObservers(message);
}

void QuestComponent::OnNotify(const EventData& message)
{
	if (std::holds_alternative<int>(message))
	{
		std::cout << "QuestComponent received message: " << std::get<int>(message) << std::endl;
	}
	else if (std::holds_alternative<std::string>(message))
	{
		std::cout << "QuestComponent received message: " << std::get<std::string>(message) << std::endl;
	}
	else if (std::holds_alternative<ComponentEventExample>(message))
	{
		std::cout << "QuestComponent received message: " << std::get<ComponentEventExample>(message).action << std::endl;
	}
}

File: BB1/Server/GameServer/Game/Entity/Component/QuestComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Google.Protobuf;
using Protocol = Google.Protobuf.Protocol;

namespace GameServer
{
    public class QuestComponent : EntityComponent
    {
        public override void Update()
        {
            Console.WriteLine("[QuestComponent] Updating quest progress...");

            int updatedValue = 100;
            Console.WriteLine($"[QuestComponent] New value: {updatedValue}");

            NotifyObservers(new EventData { Type = EventDataType.Int, Data = updatedValue });
        }

        public override void Notify(EventData message)
        {
            Console.WriteLine("[QuestComponent] Notifying observers with quest message.");
            NotifyObservers(message);
        }

        public override void OnNotify(EventData message)
        {
            switch (message.Type)
            {
                case EventDataType.Int:
                    Console.WriteLine($"[QuestComponent] Received int message: {(int)message.Data}");
                    break;
                case EventDataType.Float:
                    Console.WriteLine($"[QuestComponent] Received float message: {(float)message.Data}");
                    break;
                case EventDataType.String:
                    Console.WriteLine($"[QuestComponent] Received string message: {(string)message.Data}");
                    break;
                default:
                    Console.WriteLine("[QuestComponent] Unknown message type received.");
                    break;
            }
        }
    }
}